File: core/strategies/analyze.py
"""Unified strategy analysis aggregator.

Domain:    Strategy Analysis — Orchestrator
Contracts:
  - analyze_strategy(legs, spot, price_range, n_points, r) -> dict
Dependencies UPWARD:
  - core.strategies.payoff, core.strategies.greeks, core.strategies.prob_profit
Dependencies DOWNWARD:
  - services.strategy_service, tests
"""

from __future__ import annotations

import numpy as np

from core.strategies.greeks import net_greeks
from core.strategies.models import Leg
from core.strategies.payoff import find_breakevens, net_premium, payoff_at_expiration
from core.strategies.prob_profit import prob_profit


def analyze_strategy(
    legs: list[Leg],
    spot: float,
    *,
    price_range: tuple[float, float] | None = None,
    n_points: int = 401,
    r: float = 0.05,
) -> dict:
    """Compute payoff, breakevens, Greeks, max P&L, and probability of profit.

    Returns dict with keys: prices, pnl, breakevens, max_profit, max_loss,
    net_premium, greeks, prob_profit.
    """
    if not legs:
        raise ValueError("analyze_strategy: legs must not be empty")
    if spot <= 0:
        raise ValueError("analyze_strategy: spot must be positive")

    strikes = [leg.strike for leg in legs]
    if price_range is None:
        lo = min(min(strikes), spot) * 0.85
        hi = max(max(strikes), spot) * 1.15
    else:
        lo, hi = price_range
    lo = max(lo, 0.01)
    prices = np.linspace(lo, hi, n_points)
    pnl = payoff_at_expiration(legs, prices)

    edge_left = pnl[1] - pnl[0]
    edge_right = pnl[-1] - pnl[-2]
    long_calls = sum(leg.qty for leg in legs if leg.side == "long" and leg.option_type == "call")
    short_calls = sum(leg.qty for leg in legs if leg.side == "short" and leg.option_type == "call")
    long_puts = sum(leg.qty for leg in legs if leg.side == "long" and leg.option_type == "put")
    short_puts = sum(leg.qty for leg in legs if leg.side == "short" and leg.option_type == "put")
    upside_naked = short_calls > long_calls
    downside_naked = short_puts > long_puts

    max_profit = float(np.max(pnl))
    max_loss = float(np.min(pnl))
    if upside_naked and edge_right < 0:
        max_loss = float("-inf")
    if downside_naked and edge_left > 0:
        max_loss = float("-inf")
    if long_calls > short_calls and edge_right > 0:
        max_profit = float("inf")

    dte_max = max((leg.dte for leg in legs), default=30)
    iv_w_num = sum(leg.iv * leg.qty for leg in legs)
    iv_w_den = sum(leg.qty for leg in legs) or 1
    iv_avg = iv_w_num / iv_w_den

    return {
        "prices": prices.tolist(),
        "pnl": pnl.tolist(),
        "breakevens": find_breakevens(prices, pnl),
        "max_profit": max_profit,
        "max_loss": max_loss,
        "net_premium": net_premium(legs),
        "greeks": net_greeks(legs, spot, r=r),
        "prob_profit": prob_profit(prices, pnl, spot, iv_avg, dte_max, r=r),
    }

File: site/static/sim/analyze.js
// analyze.js — multi-leg strategy aggregation, mirroring
// core/strategies/analyze.py::analyze_strategy.
//
// Returns the SAME shape the existing browser renderer expects
// (prices, pnl, breakevens, max_profit, max_loss, net_premium, greeks,
// prob_profit), so static/components/payoff_chart.js::renderPayoff works
// unchanged. The naked-leg ±inf logic is copied VERBATIM — including the
// deliberate "long put is NOT treated as unbounded" detail.

import { bsGreeks } from './black_scholes.js';
import { payoffAtExpiration, netPremium, findBreakevens } from './payoff.js';
import { probProfit } from './stats.js';

// Banker's rounding (round half to even) to match Python's round() used by
// core/strategies/greeks.py::net_greeks.
export function round4(x) {
  if (!Number.isFinite(x)) return x;
  const f = 1e4;
  const y = x * f;
  const rb = Math.round(y);
  if (Math.abs(y - rb) === 0.5 && Math.abs(rb % 2) === 1) return (rb - 1) / f;
  return rb / f;
}

export function netGreeks(legs, spot, r = 0.05) {
  const totals = { delta: 0, gamma: 0, theta: 0, vega: 0 };
  for (const leg of legs) {
    const T = Math.max(leg.dte ?? 30, 1) / 365;
    const g = bsGreeks(spot, leg.strike, T, r, leg.iv ?? 0.25, leg.optionType);
    const scale = (leg.side === 'long' ? 1 : -1) * (leg.qty ?? 1);
    for (const k of Object.keys(totals)) {
      const v = g[k];
      if (Number.isFinite(v)) totals[k] += scale * v;
    }
  }
  const out = {};
  for (const k of Object.keys(totals)) out[k] = round4(totals[k]);
  return out;
}

export function analyzeStrategy(legs, spot, opts = {}) {
  const { priceRange = null, nPoints = 401, r = 0.05 } = opts;
  if (!legs || legs.length === 0) throw new Error('analyzeStrategy: legs must not be empty');
  if (!(spot > 0)) throw new Error('analyzeStrategy: spot must be positive');

  const strikes = legs.map((l) => l.strike);
  let lo, hi;
  if (priceRange) {
    [lo, hi] = priceRange;
  } else {
    lo = Math.min(Math.min(...strikes), spot) * 0.85;
    hi = Math.max(Math.max(...strikes), spot) * 1.15;
  }
  lo = Math.max(lo, 0.01);

  const prices = [];
  for (let i = 0; i < nPoints; i++) prices.push(lo + ((hi - lo) * i) / (nPoints - 1));

  const pnl = payoffAtExpiration(legs, prices);

  const edgeLeft = pnl[1] - pnl[0];
  const edgeRight = pnl[pnl.length - 1] - pnl[pnl.length - 2];

  let longCalls = 0, shortCalls = 0, longPuts = 0, shortPuts = 0;
  for (const l of legs) {
    const q = l.qty ?? 1;
    if (l.optionType === 'call') {
      if (l.side === 'long') longCalls += q;
      else shortCalls += q;
    } else {
      if (l.side === 'long') longPuts += q;
      else shortPuts += q;
    }
  }
  const upsideNaked = shortCalls > longCalls;
  const downsideNaked = shortPuts > longPuts;

  let maxProfit = Math.max(...pnl);
  let maxLoss = Math.min(...pnl);
  if (upsideNaked && edgeRight < 0) maxLoss = -Infinity;
  if (downsideNaked && edgeLeft > 0) maxLoss = -Infinity;
  if (longCalls > shortCalls && edgeRight > 0) maxProfit = Infinity;

  const dteMax = Math.max(...legs.map((l) => l.dte ?? 30), 30);
  let ivNum = 0, ivDen = 0;
  for (const l of legs) {
    ivNum += (l.iv ?? 0.25) * (l.qty ?? 1);
    ivDen += l.qty ?? 1;
  }
  const ivAvg = ivDen ? ivNum / ivDen : 0;

  return {
    prices,
    pnl,
    breakevens: findBreakevens(prices, pnl),
    maxProfit,
    maxLoss,
    netPremium: netPremium(legs),
    greeks: netGreeks(legs, spot, r),
    probProfit: probProfit(prices, pnl, spot, ivAvg, dteMax, r),
  };
}
